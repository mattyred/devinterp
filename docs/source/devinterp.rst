devinterp package
=================

Subpackages
-----------

.. toctree::
   :maxdepth: 4

   devinterp.backends
   devinterp.mechinterp
   devinterp.optim
   devinterp.slt

Submodules
----------

devinterp.test\_utils module
----------------------------

.. automodule:: devinterp.test_utils
   :members:
   :undoc-members:
   :show-inheritance:

devinterp.utils module
----------------------

.. automodule:: devinterp.utils
   :members:
   :undoc-members:
   :show-inheritance:

devinterp.vis\_utils module
---------------------------

.. automodule:: devinterp.vis_utils
   :members:
   :undoc-members:
   :show-inheritance:

Module contents
---------------

.. automodule:: devinterp
   :members:
   :undoc-members:
   :show-inheritance:
