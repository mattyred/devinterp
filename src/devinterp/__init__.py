__all__ = ["devinterp.slt", "devinterp.optim"]
