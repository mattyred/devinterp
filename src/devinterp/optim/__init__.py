from devinterp.optim.sgld import SGLD as SGLD
from devinterp.optim.sgmcmc import SGMCMC as SGMCMC
from devinterp.optim.sgmcmc import SamplingMethodLiteral as SamplingMethodLiteral
from devinterp.optim.sgnht import SGNHT as SGNHT
