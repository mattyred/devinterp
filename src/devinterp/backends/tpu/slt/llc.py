from typing import Union

import numpy as np
import torch
import torch_xla.core.xla_model as xm
from devinterp.slt.callback import SamplerCallback


class LLCEstimator(SamplerCallback):
    r"""
    Callback for estimating the Local Learning Coefficient (LLC) in a rolling fashion during a sampling process.
    It calculates the LLC based on the average loss across draws for each chain:

    $$LLC = \textrm{T} * (\textrm{avg_loss} - \textrm{init_loss})$$

    For use with :func:`devinterp.slt.sampler.sample`.

    Note:
        - `init_loss` gets set inside :func:`devinterp.slt.sample()`. It can be passed as an argument to that function,
        and if not passed will be the average loss of the supplied model over `num_chains` batches.

    :param num_draws: Number of samples to draw (should be identical to :python:`num_draws` passed to :python:`devinterp.slt.sampler.sample`)
    :type num_draws: int
    :param num_chains: Number of chains to run (should be identical to :python:`num_chains` passed to :python:`devinterp.slt.sampler.sample`)
    :type num_chains: int
    :param nbeta: nbeta, float (default: 1., set by sample() to utils.optimal_nbeta(dataloader)=len(batch_size)/np.log(len(batch_size)))
    :type nbeta: int
    :param device: Device to perform computations on, e.g., 'cpu' or 'cuda'.
    :type device: str | torch.device, optional
    """

    def __init__(
        self,
        num_chains: int,
        num_draws: int,
        nbeta: float,
        init_loss: torch.Tensor = None,
        device: Union[torch.device, str] = "cpu",
        eval_field: str = "loss",
    ):
        self.num_chains = num_chains
        self.num_draws = num_draws
        self.loss = torch.zeros(1, dtype=torch.float32).to(device)
        self.sq_loss = torch.zeros(1, dtype=torch.float32).to(device)
        self.init_loss = init_loss or torch.zeros(1, dtype=torch.float32).to(device)
        self.nbeta = torch.tensor(nbeta, dtype=torch.float32).to(device)
        self.device = device
        self.count = 0.0
        self.eval_field = eval_field

    def update(self, chain: int, draw: int, loss: torch.Tensor):
        self.loss += loss
        self.sq_loss += loss**2
        self.count += 1.0

    def finalize(self):
        scale = 1.0 / (self.count * xm.xrt_world_size())
        self.loss = xm.all_reduce(xm.REDUCE_SUM, self.loss, scale=scale)
        self.sq_loss = xm.all_reduce(xm.REDUCE_SUM, self.sq_loss, scale=scale)

        self.llc = self.nbeta * (self.loss - self.init_loss)

    def sample(self):
        """
        :returns: A dict :python:`{"llc/mean": llc_mean, "llc/std": llc_std, "llc-chain/{i}": llc_trace_per_chain, "loss/trace": loss_trace_per_chain}`. (Only after running :python:`devinterp.slt.sampler.sample(..., [llc_estimator_instance], ...)`).
        """
        return {
            "init_loss": self.init_loss.cpu().item(),
            "llc": self.llc.cpu().item(),
            "loss": self.loss.cpu().item(),
            "sq_loss": self.sq_loss.cpu().item(),
        }

    def __call__(self, chain: int, draw: int, *args, **kwargs):
        self.update(chain, draw, kwargs[self.eval_field])



class OnlineLLCEstimator(SamplerCallback):
    """
    Callback for estimating the Local Learning Coefficient (LLC) in an online fashion during a sampling process.
    It calculates LLCs using the same formula as :func:`devinterp.slt.llc.LLCEstimator`, but continuously and including means and std across draws (as opposed to just across chains).
    For use with :func:`devinterp.slt.sampler.sample`.

    Note:
        - `init_loss` gets set inside :func:`devinterp.slt.sample()`. It can be passed as an argument to that function,
        and if not passed will be the average loss of the supplied model over `num_chains` batches.
    :param num_draws: Number of samples to draw (should be identical to :python:`num_draws` passed to :python:`devinterp.slt.sampler.sample`)
    :type num_draws: int
    :param num_chains: Number of chains to run (should be identical to :python:`num_chains` passed to :python:`devinterp.slt.sampler.sample`)
    :type num_chains: int
    :param nbeta: nbeta, float (default: 1., set by sample() to utils.optimal_nbeta(dataloader)=len(batch_size)/np.log(len(batch_size)))
    :type nbeta: int
    :param device: Device to perform computations on, e.g., 'cpu' or 'cuda'. Default is 'cpu'
    :type device: str | torch.device, optional
    """

    def __init__(
        self,
        num_chains: int,
        num_draws: int,
        nbeta: float,
        init_loss,
        device="cpu",
        eval_field="loss",
    ):
        self.num_chains = num_chains
        self.num_draws = num_draws
        self.init_loss = init_loss  # gets set by devinterp.slt.sample()

        self.losses = [[] for _ in range(num_chains)]
        self.llcs = np.zeros((num_chains, num_draws))
        self.nbeta = nbeta
        self.device = device
        self.eval_field = eval_field

    def update(self, chain: int, draw: int, loss: float):
        self.losses[chain].append(loss)

    def finalize(self):
        self.losses = torch.tensor(self.losses).cpu().numpy()
        cum_loss_avgs = np.cumsum(np.mean(self.losses, axis=0), axis=0) / np.arange(
            1, self.num_draws + 1
        )
        self.llcs = self.nbeta * (cum_loss_avgs - self.init_loss)

    def sample(self):
        """
        :returns: A dict :python:`{"llc/means": llc_means, "llc/stds": llc_stds, "llc/trace": llc_trace_per_chain, "loss/trace": loss_trace_per_chain}`. (Only after running :python:`devinterp.slt.sampler.sample(..., [llc_estimator_instance], ...)`).
        """
        return {
            "init_loss": self.init_loss,
            "llc": self.llcs[-1],
            "loss": self.losses[-1],
            "llcs": self.llcs,
            "losses": self.losses,
        }

    def __call__(self, chain: int, draw: int, *args, **kwargs):
        self.update(chain, draw, kwargs[self.eval_field])


def llc_estimator_factory(
    num_chains: int,
    num_draws: int,
    nbeta: float,
    init_loss,
    device="cpu",
    eval_field: str = "loss",
    online: bool = False,
    per_token: bool = False,
) -> SamplerCallback:
    if online and per_token:
        raise ValueError("Cannot have both online and per_token set to True")
    elif online:
        return OnlineLLCEstimator(
            num_chains=num_chains,
            num_draws=num_draws,
            nbeta=nbeta,
            init_loss=init_loss,
            device=device,
            eval_field=eval_field,
        )
    # elif per_token:
    #     return PerTokenLLCEstimator(
    #         num_chains=num_chains,
    #         num_draws=num_draws,
    #         nbeta=nbeta,
    #         init_loss=init_loss,
    #         device=device,
    #         eval_field=eval_field,
    #     )
    else:
        return LLCEstimator(
            num_chains=num_chains,
            num_draws=num_draws,
            nbeta=nbeta,
            device=device,
            init_loss=init_loss,
            eval_field=eval_field,
        )
