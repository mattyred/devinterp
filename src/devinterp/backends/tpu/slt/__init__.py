from devinterp.backends.tpu.slt.llc import *
from devinterp.backends.tpu.slt.sampler import *
