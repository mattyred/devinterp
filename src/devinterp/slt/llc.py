from typing import Union, Optional
import torch
from devinterp.slt.callback import SamplerCallback
import warnings

# TPU support
try:
    import torch_xla.core.xla_model as xm
    USING_XLA = True
except ImportError:
    USING_XLA = False

class LLCEstimator(SamplerCallback):
    r"""
    Callback for estimating the Local Learning Coefficient (LLC) in a rolling fashion during a sampling process.
    It calculates the LLC based on the average loss across draws for each chain:
    
    $$LLC = \textrm{T} * (\textrm{avg_loss} - \textrm{init_loss})$$

    For use with :func:`devinterp.slt.sampler.sample`.
    
    Note: 
        - `init_loss` gets set inside :func:`devinterp.slt.get_results()`. It can be passed as an argument to that function, 
        and if not passed will be the average loss of the supplied model over `num_chains` batches.

    :param num_draws: Number of samples to draw (should be identical to :python:`num_draws` passed to :python:`devinterp.slt.sampler.sample`)
    :type num_draws: int
    :param num_chains: Number of chains to run (should be identical to :python:`num_chains` passed to :python:`devinterp.slt.sampler.sample`)
    :type num_chains: int
    :param nbeta: Reparameterized inverse temperature
    :type nbeta: float
    :param device: Device to perform computations on, e.g., 'cpu' or 'cuda'.
    :type device: str | torch.device, optional
    """
    def __init__(
        self,
        num_chains: int,
        num_draws: int,
        init_loss: torch.Tensor,
        eval_field: str = "loss",
        device: Union[torch.device, str] = "cpu",
        nbeta: float = None,
        temperature: Optional[float] = None, # Temperature is deprecated
    ):
        self.num_chains = num_chains
        self.num_draws = num_draws

        self.loss = torch.zeros((num_chains, num_draws), dtype=torch.float32).to(device)
        self.init_loss = init_loss

        assert nbeta is not None or temperature is not None, "Please provide a value for nbeta."
        if nbeta is None and temperature is not None:
            nbeta = temperature
            warnings.warn("Temperature is deprecated. Please use nbeta instead.")
        self.nbeta = torch.tensor(nbeta, dtype=torch.float32).to(device)
        self.eval_field = eval_field
        self.device = device
        self.count = 0.0

    def update(self, chain: int, draw: int, loss: torch.Tensor, **kwargs):
        """
        Called at each step of sampling.

        Order of operations: 
        1. update() is called at each draw.
        2. finalize() is called at the end of sampling.
        3. get_results() is called after finalize().

        :param draw: The current draw index.
        :type draw: int
        :param loss: The loss at the current draw.
        :type loss: float
        :param chain: The chain index. Note that this is not used here at all.
        :type chain: int
        """
        self.loss[chain, draw] = loss
        self.count += 1.0

    def finalize(self):
        """Called once at the end of sampling."""

        # If using TPUs, aggregate the loss across all devices.
        if USING_XLA:
            if self.count == 0:
                self.count = 1
            scale = 1.0 / (self.count * xm.xrt_world_size())
            self.loss = xm.all_reduce(xm.REDUCE_SUM, self.loss, scale=scale)

        avg_loss = self.loss.mean(axis=1) # [num_chains]. Average loss across draws for each chain.

        self.llc_per_chain = self.nbeta * (avg_loss - self.init_loss) # [num_chains]. A single LLC value for each chain.
        self.sq_loss = torch.square(self.loss)
        self.llc_mean = self.llc_per_chain.mean() # Mean LLC across chains.
        self.llc_std = self.llc_per_chain.std() # Standard deviation of LLC across chains.

    def get_results(self):
        """
        Method to return results. Called after finalize().

        :returns: A dict :python:`{"llc/mean": llc_mean, "llc/std": llc_std, "llc-chain/{i}": llc_trace_per_chain, "loss/trace": loss_trace_per_chain}`. (Only after running :python:`devinterp.slt.sampler.sample(..., [llc_estimator_instance], ...)`).
        """
        return {
            "llc/mean": self.llc_mean.cpu().numpy().item(),
            "llc/std": self.llc_std.cpu().numpy().item(),
            **{
                f"llc-chain/{i}": self.llc_per_chain[i].cpu().numpy().item()
                for i in range(self.num_chains)
            },
            "loss/trace": self.loss.cpu().numpy(),
            "loss/init_loss": self.init_loss,
            "loss/sq_loss": self.sq_loss.cpu().numpy(),
        }

    def __call__(self, chain: int, draw: int, **kwargs):
        """Calls self.update."""
        self.update(chain, draw, kwargs[self.eval_field])


class OnlineLLCEstimator(SamplerCallback):
    """
    Callback for estimating the Local Learning Coefficient (LLC) in an online fashion during a sampling process.
    It calculates LLCs using the same formula as :func:`devinterp.slt.llc.LLCEstimator`, but continuously and including means and std across draws (as opposed to just across chains).
    For use with :func:`devinterp.slt.sampler.sample`.
    
    Note: 
        - `init_loss` gets set inside :func:`devinterp.slt.get_results()`. It should be passed as an argument to that function.
    :param num_draws: Number of samples to draw (should be identical to :python:`num_draws` passed to :python:`devinterp.slt.sampler.sample`)
    :type num_draws: int
    :param num_chains: Number of chains to run (should be identical to :python:`num_chains` passed to :python:`devinterp.slt.sampler.sample`)
    :type num_chains: int
    :param nbeta: Inverse reparameterized temperature
    :type nbeta: float
    :param device: Device to perform computations on, e.g., 'cpu' or 'cuda'. Default is 'cpu'
    :type device: str | torch.device, optional
    """

    def __init__(
        self, 
        num_chains: int, 
        num_draws: int, 
        init_loss: torch.Tensor,
        device="cpu",
        eval_field="loss",
        nbeta: float = None,
        temperature: Optional[float] = None, # Temperature is deprecated
    ):
        self.num_draws = num_draws
        self.init_loss = init_loss

        self.losses = torch.zeros((num_chains, num_draws)).to(device)
        self.llcs = torch.zeros((num_chains, num_draws)).to(device)

        assert nbeta is not None or temperature is not None, "Please provide a value for nbeta."
        if nbeta is None and temperature is not None:
            nbeta = temperature
            warnings.warn("Temperature is deprecated. Please use nbeta instead.")
        self.nbeta = torch.tensor(nbeta, dtype=torch.float32).to(device)
        
        self.device = device
        self.eval_field = eval_field

        self.llc_means = torch.tensor(num_draws, dtype=torch.float32).to(device)
        self.llc_stds = torch.tensor(num_draws, dtype=torch.float32).to(device)
        self.moving_avg_llcs = torch.zeros(
            (num_chains, num_draws), dtype=torch.float32
        ).to(device)
        
    def update(self, chain: int, draw: int, loss: float):
        self.losses[chain, draw] = loss
        self.llcs[chain, draw] = self.nbeta * (loss - self.init_loss)
        if draw == 0:
            self.moving_avg_llcs[chain, draw] = self.nbeta * (
                loss - self.init_loss
            )
        else:
            t = draw + 1
            prev_moving_avg = self.moving_avg_llcs[chain, draw - 1]
            self.moving_avg_llcs[chain, draw] = (1 / t) * (
                (t - 1) * prev_moving_avg + self.nbeta * (loss - self.init_loss)
            )

    def finalize(self):
        self.llc_means = self.llcs.mean(dim=0)
        self.llc_stds = self.llcs.std(dim=0)

    def get_results(self):
        """    
        :returns: A dict :python:`{"llc/means": llc_means, "llc/stds": llc_stds, "llc/trace": llc_trace_per_chain, "loss/trace": loss_trace_per_chain}`. (Only after running :python:`devinterp.slt.sampler.sample(..., [llc_estimator_instance], ...)`).
        """
        return {
            "llc/means": self.llc_means.cpu().numpy(),
            "llc/moving_avg": self.moving_avg_llcs.cpu().numpy(),
            "llc/stds": self.llc_stds.cpu().numpy(),
            "llc/trace": self.llcs.cpu().numpy(),
            "loss/trace": self.losses.cpu().numpy(),
        }

    def __call__(self, chain: int, draw: int, **kwargs):
        self.update(chain, draw, kwargs[self.eval_field])
