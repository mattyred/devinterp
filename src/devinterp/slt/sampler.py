import itertools
import warnings
from copy import deepcopy
from typing import Callable, Dict, List, Literal, Optional, Type, Union

import torch
from torch import nn
from torch.multiprocessing import cpu_count, get_context
from torch.utils.data import DataLoader
from tqdm import tqdm

from devinterp.backends.default.slt.llc import LLCEstimator, OnlineLLCEstimator
from devinterp.backends.default.slt.mala import MalaAcceptanceRate
from devinterp.backends.default.slt.norms import NoiseNorm
from devinterp.optim.sgld import SGLD
from devinterp.slt.callback import SamplerCallback, validate_callbacks
from devinterp.utils import (
    USE_TPU_BACKEND,
    EvaluateFn,
    call_with,
    get_init_loss_multi_batch,
    optimal_temperature,
    prepare_input,
    split_results,
)

if USE_TPU_BACKEND:
    from devinterp.backends.tpu.slt.sampler import sample
else:
    from devinterp.backends.default.slt.sampler import sample


def estimate_learning_coeff_with_summary(
    model: torch.nn.Module,
    loader: DataLoader,
    callbacks: List[Callable] = [],
    evaluate: Optional[EvaluateFn] = None,
    sampling_method: Type[torch.optim.Optimizer] = SGLD,
    optimizer_kwargs: Optional[Dict[str, Union[float, Literal["adaptive"]]]] = None,
    num_draws: int = 100,
    num_chains: int = 10,
    num_burnin_steps: int = 0,
    num_steps_bw_draws: int = 1,
    init_loss: float = None,
    grad_accum_steps: int = 1,
    cores: int = 1,
    seed: Optional[Union[int, List[int]]] = None,
    device: Union[torch.device, str] = torch.device("cpu"),
    verbose: bool = True,
    optimize_over_per_model_param: Optional[Dict[str, List[bool]]] = None,
    online: bool = False,
) -> dict:
    optimizer_kwargs.setdefault("temperature", optimal_temperature(loader))
    if not init_loss:
        init_loss = get_init_loss_multi_batch(
            loader, num_chains, model, evaluate, device
        )
        # alternative: init_loss = get_init_loss_full_batch(loader, model, evaluate, device)
        # alternative: init_loss = get_init_loss_one_batch(loader, model, evaluate, device)
    if online:
        llc_estimator = OnlineLLCEstimator(
            num_chains, num_draws, optimizer_kwargs["temperature"], device=device
        )
    else:
        llc_estimator = LLCEstimator(
            num_chains, num_draws, optimizer_kwargs["temperature"], device=device
        )

    callbacks = [llc_estimator, *callbacks]

    sample(
        model=model,
        loader=loader,
        evaluate=evaluate,
        sampling_method=sampling_method,
        optimizer_kwargs=optimizer_kwargs,
        num_draws=num_draws,
        num_chains=num_chains,
        num_burnin_steps=num_burnin_steps,
        num_steps_bw_draws=num_steps_bw_draws,
        grad_accum_steps=grad_accum_steps,
        cores=cores,
        seed=seed,
        device=device,
        verbose=verbose,
        callbacks=callbacks,
        init_loss=init_loss,
        optimize_over_per_model_param=optimize_over_per_model_param,
    )

    results = {}

    for callback in callbacks:
        if hasattr(callback, "sample"):
            results.update(callback.sample())

    return results


def estimate_learning_coeff(
    model: torch.nn.Module,
    loader: DataLoader,
    callbacks: List[Callable] = [],
    evaluate: Optional[EvaluateFn] = None,
    sampling_method: Type[torch.optim.Optimizer] = SGLD,
    optimizer_kwargs: Optional[Dict[str, Union[float, Literal["adaptive"]]]] = None,
    num_draws: int = 100,
    num_chains: int = 10,
    num_burnin_steps: int = 0,
    num_steps_bw_draws: int = 1,
    init_loss: float = None,
    grad_accum_steps: int = 1,
    cores: int = 1,
    seed: Optional[Union[int, List[int]]] = None,
    device: Union[torch.device, str] = torch.device("cpu"),
    verbose: bool = True,
    optimize_over_per_model_param: Optional[Dict[str, List[bool]]] = None
) -> float:
    return estimate_learning_coeff_with_summary(
        model=model,
        loader=loader,
        evaluate=evaluate,
        sampling_method=sampling_method,
        optimizer_kwargs=optimizer_kwargs,
        num_draws=num_draws,
        num_chains=num_chains,
        num_burnin_steps=num_burnin_steps,
        num_steps_bw_draws=num_steps_bw_draws,
        grad_accum_steps=grad_accum_steps,
        cores=cores,
        seed=seed,
        device=device,
        verbose=verbose,
        callbacks=callbacks,
        online=False,
        init_loss=init_loss,
        optimize_over_per_model_param=optimize_over_per_model_param,
    )["llc/mean"]
