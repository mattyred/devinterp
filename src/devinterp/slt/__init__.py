from devinterp.slt.sampler import *