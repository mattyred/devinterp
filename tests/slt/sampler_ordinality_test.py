import numpy as np
import pytest
import torch
import torch.nn as nn
import torch.nn.functional as F
import platform
from devinterp.optim import SGLD, SGMCMC
from devinterp.slt.llc import LLCEstimator
from devinterp.slt.sampler import sample
from devinterp.utils import default_nbeta, evaluate_mse, get_init_loss_multi_batch
from torch.utils.data import DataLoader, TensorDataset

# Test configuration constants
SNAPSHOT_DRAWS = 5
FULL_SAMPLING_DRAWS = 1000
NUM_CHAINS = 5
RANDOM_SEED = 42


def _do_ordinality_sampling(
    model, train_dataloader, sampling_method, lr, sample_points, num_draws
):
    """Perform MCMC sampling to estimate LLC for ordinality test.

    Args:
        model: Model to sample from
        train_dataloader: DataLoader for training data
        sampling_method: SGLD or SGMCMC sampling method
        lr: Learning rate
        sample_points: List of sample points to test
        num_draws: Number of MCMC draws to perform

    Returns:
        List of LLC means for each sample point
    """
    torch.manual_seed(RANDOM_SEED)

    llcs = []
    for sample_point in sample_points:
        model.weights = nn.Parameter(
            torch.tensor(sample_point, dtype=torch.float32, requires_grad=True)
        )
        init_loss = get_init_loss_multi_batch(
            train_dataloader, NUM_CHAINS, model, evaluate_mse, device="cpu"
        )
        llc_estimator = LLCEstimator(
            num_chains=NUM_CHAINS,
            num_draws=num_draws,
            nbeta=default_nbeta(train_dataloader),
            init_loss=init_loss,
        )

        sample(
            model,
            train_dataloader,
            evaluate=evaluate_mse,
            sampling_method_kwargs=dict(
                lr=lr,
                bounding_box_size=0.5,
                nbeta=default_nbeta(train_dataloader),
                # to prevent accidental movement from [1, 0, ...] to origin
            ),
            sampling_method=sampling_method,
            num_chains=NUM_CHAINS,
            num_draws=num_draws,
            callbacks=[llc_estimator],
            verbose=False,
            seed=RANDOM_SEED,
        )
        llcs.append(llc_estimator.get_results()["llc/mean"])

    return llcs


def _test_ordinality_accuracy(
    model, train_dataloader, sampling_method, lr, sample_points, model_name, dim
):
    """Test LLC ordinality with full sampling."""
    llcs = _do_ordinality_sampling(
        model,
        train_dataloader,
        sampling_method,
        lr,
        sample_points,
        num_draws=FULL_SAMPLING_DRAWS,
    )

    assert (
        np.diff(llcs) >= 0
    ).all(), f"Ordinality not preserved for sampler {sampling_method} on {dim}-d {model_name}: llcs {llcs} are not in ascending order."


@pytest.mark.skipif(
    platform.machine() != "x86_64",
    reason=f"Differences in results between ARM and x86_64. Your arch is {platform.machine()}",
)
@pytest.mark.parametrize("sampling_method", [SGLD, SGMCMC.sgld])
@pytest.mark.parametrize("model_name", ["Polynomial", "LinePlusDot"])
@pytest.mark.parametrize("dim", [2, 10])
def test_linedot_normal_crossing(
    generated_linedot_normalcrossing_dataset,
    sampling_method,
    model_name,
    dim,
    request,
    snapshot,
    is_snapshot_update,
):
    torch.manual_seed(RANDOM_SEED)
    Model = request.getfixturevalue(model_name)
    if model_name == "Polynomial":
        model = Model([2 for _ in range(dim)])
    else:
        model = Model(dim)
    train_dataloader, _, _, _ = generated_linedot_normalcrossing_dataset
    lr = (
        0.0001 / dim
    )  # to account for smaller steps in higher D. might not work well for SGNHT?

    sample_points = [
        [0.0 for _ in range(dim)],
        [0.0 if i == dim - 1 else 1.0 for i in range(dim)],
    ]

    # Run snapshot test first
    llcs = _do_ordinality_sampling(
        model,
        train_dataloader,
        sampling_method,
        lr,
        sample_points,
        num_draws=SNAPSHOT_DRAWS,
    )

    # Run verification test when updating snapshots
    if is_snapshot_update:
        _test_ordinality_accuracy(
            model, train_dataloader, sampling_method, lr, sample_points, model_name, dim
        )

    # Test against snapshot
    assert llcs == snapshot
