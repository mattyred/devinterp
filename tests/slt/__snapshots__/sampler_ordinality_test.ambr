# serializer version: 1
# name: test_linedot_normal_crossing[10-LinePlusDot-SGLD]
  list([
    8.628656473774754e-07,
    125492.046875,
  ])
# ---
# name: test_linedot_normal_crossing[10-LinePlusDot-sgld]
  list([
    8.628656473774754e-07,
    125492.046875,
  ])
# ---
# name: test_linedot_normal_crossing[10-Polynomial-SGLD]
  list([
    0.0,
    -5.7380566431675106e-05,
  ])
# ---
# name: test_linedot_normal_crossing[10-Polynomial-sgld]
  list([
    0.0,
    -5.7380566431675106e-05,
  ])
# ---
# name: test_linedot_normal_crossing[2-LinePlusDot-SGLD]
  list([
    2.804313453452778e-06,
    0.05154026299715042,
  ])
# ---
# name: test_linedot_normal_crossing[2-LinePlusDot-sgld]
  list([
    2.804313453452778e-06,
    0.05154026299715042,
  ])
# ---
# name: test_linedot_normal_crossing[2-Polynomial-SGLD]
  list([
    0.0,
    -0.0008089365437626839,
  ])
# ---
# name: test_linedot_normal_crossing[2-Polynomial-sgld]
  list([
    0.0,
    -0.0008089365437626839,
  ])
# ---
